File: mobile/src/api/entities.ts
import { apiClient } from "./client";
import { makeResource } from "./resource";
import type {
  ChargeOptionResult, ChargeRule, Customer, DebtSummary, Driver, DriverLedgerEntry, DriverSettlement,
  Load, Place, Product, ProductChargeOption, Trip, TripDebtSummary, TripExpense, Vehicle, VehicleExpense,
} from "@/types";

export const vehiclesApi = makeResource<Vehicle>("/vehicles");
export const customersApi = makeResource<Customer>("/customers");
export const placesApi = makeResource<Place>("/places");
export const vehicleExpensesApi = makeResource<VehicleExpense>("/vehicle-expenses");

const productBase = makeResource<Product>("/products");
export const productsApi = {
  ...productBase,
  chargeOptions: {
    list: (productId: string) =>
      apiClient.get<ProductChargeOption[]>(`/products/${productId}/charge-options`).then((r) => r.data),
    batchCreate: (productId: string, options: { charge_type: string; kg_variant?: number; rate: number }[]) =>
      apiClient
        .post<ProductChargeOption[]>(`/products/${productId}/charge-options/batch`, { options })
        .then((r) => r.data),
    update: (optionId: string, payload: { charge_type: string; kg_variant?: number; rate: number }) =>
      apiClient.patch<ProductChargeOption>(`/products/charge-options/${optionId}`, payload).then((r) => r.data),
    remove: (optionId: string) => apiClient.delete(`/products/charge-options/${optionId}`),
  },
};

export const chargeRulesApi = {
  list: (params?: { product_id?: string; customer_id?: string }) =>
    apiClient.get<ChargeRule[]>("/charge-rules/", { params }).then((r) => r.data),
  batchCreate: (payload: {
    scope: "customer" | "route";
    product_id: string;
    customer_id?: string;
    origin_place_id?: string;
    destination_place_id?: string;
    options: { charge_type: string; kg_variant?: number; rate: number }[];
  }) => apiClient.post<ChargeRule[]>("/charge-rules/batch", payload).then((r) => r.data),
  update: (id: string, payload: Partial<ChargeRule>) =>
    apiClient.patch<ChargeRule>(`/charge-rules/${id}`, payload).then((r) => r.data),
  remove: (id: string) => apiClient.delete(`/charge-rules/${id}`),
};

const driverBase = makeResource<Driver>("/drivers");
export const driversApi = {
  ...driverBase,
  debtSummary: (id: string) =>
    apiClient.get<DebtSummary>(`/drivers/${id}/debt-summary`).then((r) => r.data),
  ledger: (id: string) =>
    apiClient.get<DriverLedgerEntry[]>(`/drivers/${id}/ledger`).then((r) => r.data),
  recordSettlement: (id: string, payload: { driver_id: string; amount: number; date: string; note?: string; trip_id?: string }) =>
    apiClient.post<DriverSettlement>(`/drivers/${id}/settlements`, payload).then((r) => r.data),
};

const tripBase = makeResource<Trip>("/trips");
export const tripsApi = {
  ...tripBase,
  debtSummary: (id: string) =>
    apiClient.get<TripDebtSummary>(`/trips/${id}/debt-summary`).then((r) => r.data),
  complete: (id: string) =>
    apiClient.post<Trip>(`/trips/${id}/complete`).then((r) => r.data),
};

const loadBase = makeResource<Load>("/loads");
export const loadsApi = {
  ...loadBase,
  chargeOptions: (params: {
    product_id: string; customer_id?: string;
    origin_place_id?: string; destination_place_id?: string;
  }) => apiClient.get<ChargeOptionResult[]>(`/loads/charge-options`, { params }).then((r) => r.data),
  attachToTrip: (loadId: string, tripId: string) =>
    apiClient.patch<Load>(`/loads/${loadId}/attach-to-trip`, null, { params: { trip_id: tripId } }).then((r) => r.data),
  collect: (loadId: string, payload: { discount: number; wages?: number; commission_loading?: number; commission_unloading?: number }) =>
    apiClient.post<Load>(`/loads/${loadId}/collect`, payload).then((r) => r.data),
};

export const tripExpensesApi = {
  list: (tripId: string) =>
    apiClient.get<TripExpense[]>(`/trip-expenses/`, { params: { trip_id: tripId } }).then((r) => r.data),
  create: (payload: Partial<TripExpense>) =>
    apiClient.post<TripExpense>(`/trip-expenses/`, payload).then((r) => r.data),
  update: (id: string, payload: Partial<TripExpense>) =>
    apiClient.patch<TripExpense>(`/trip-expenses/${id}`, payload).then((r) => r.data),
  remove: (id: string) => apiClient.delete(`/trip-expenses/${id}`),
};

export const reportsApi = {
  summary: (params: { date_from: string; date_to: string; driver_id?: string; vehicle_id?: string; customer_id?: string }) =>
    apiClient.get(`/reports/summary`, { params }).then((r) => r.data),
};

File: backend/app/schemas/entities.py
"""
Pydantic request/response schemas mirroring the Supabase tables.
Naming convention: <Entity>Base (shared fields) -> <Entity>Create (input) -> <Entity> (output, has id/timestamps).
"""
from datetime import date, datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ---------- Enums ----------

class ChargeType(str, Enum):
    quantity = "quantity"
    kg = "kg"
    bulk = "bulk"
    custom = "custom"


class LoadStatus(str, Enum):
    pending = "pending"
    in_transit = "in_transit"
    delivered = "delivered"
    collected = "collected"


class TripStatus(str, Enum):
    ongoing = "ongoing"
    completed = "completed"


class ChargeRuleScope(str, Enum):
    customer = "customer"        # customer + product
    route = "route"              # origin place + destination place + product


class TripExpenseCategory(str, Enum):
    fuel = "fuel"
    toll = "toll"
    driver_wage = "driver_wage"
    other = "other"


# ---------- Vehicle ----------

class VehicleBase(BaseModel):
    number: str
    type: Optional[str] = None
    capacity: Optional[float] = None
    rc_expiry: Optional[date] = None
    insurance_expiry: Optional[date] = None
    permit_expiry: Optional[date] = None


class VehicleCreate(VehicleBase):
    pass


class Vehicle(VehicleBase):
    id: UUID
    created_at: datetime


# ---------- Driver ----------

class DriverBase(BaseModel):
    name: str
    phone: Optional[str] = None
    license_number: Optional[str] = None
    license_expiry: Optional[date] = None


class DriverCreate(DriverBase):
    pass


class Driver(DriverBase):
    id: UUID
    created_at: datetime
    debt: float = 0.0  # computed, not stored — populated by the service layer


# ---------- Customer ----------

class CustomerBase(BaseModel):
    name: str
    phone: Optional[str] = None
    place_id: Optional[UUID] = None


class CustomerCreate(CustomerBase):
    pass


class Customer(CustomerBase):
    id: UUID
    created_at: datetime


# ---------- Product ----------

class ProductBase(BaseModel):
    name: str


class ProductCreate(ProductBase):
    pass


class Product(ProductBase):
    id: UUID
    created_at: datetime


class ProductChargeOptionBase(BaseModel):
    charge_type: ChargeType
    kg_variant: Optional[float] = None  # e.g. 25 / 50 / 100 — only used when charge_type == kg
    rate: float


class ProductChargeOptionCreate(ProductChargeOptionBase):
    pass


class ProductChargeOptionBatchCreate(BaseModel):
    """Lets the product-creation form submit Quantity + several KG brackets + Bulk + Custom in one call."""
    options: list[ProductChargeOptionCreate]


class ProductChargeOption(ProductChargeOptionBase):
    id: UUID
    product_id: UUID
    created_at: datetime


# ---------- Place ----------

class PlaceBase(BaseModel):
    name: str
    region: Optional[str] = None


class PlaceCreate(PlaceBase):
    pass


class Place(PlaceBase):
    id: UUID
    created_at: datetime


# ---------- Charge Rule ----------
# A rule is always scoped to a product now (rates can vary by product on the
# same route or for the same customer). Each rule row is ONE charge type/KG
# bracket; ChargeRuleBatchCreate lets the UI submit several at once (e.g.
# Quantity + 25kg + 50kg + Bulk for one customer in a single form).

class ChargeRuleBase(BaseModel):
    scope: ChargeRuleScope
    product_id: UUID
    customer_id: Optional[UUID] = None            # required when scope == customer
    origin_place_id: Optional[UUID] = None        # required when scope == route
    destination_place_id: Optional[UUID] = None   # required when scope == route


class ChargeOptionInput(BaseModel):
    """One charge-type entry within a batch submission (product creation or charge-rule creation)."""
    charge_type: ChargeType
    kg_variant: Optional[float] = None
    rate: float


class ChargeRuleCreate(ChargeRuleBase):
    charge_type: ChargeType
    kg_variant: Optional[float] = None
    rate: float


class ChargeRuleBatchCreate(ChargeRuleBase):
    options: list[ChargeOptionInput]


class ChargeRule(ChargeRuleBase):
    id: UUID
    charge_type: ChargeType
    kg_variant: Optional[float] = None
    rate: float
    created_at: datetime


class ChargeOptionResult(BaseModel):
    """
    A single selectable charge option shown to the user at load-creation time.
    Unlike the old single-best resolve_charge, ALL matching options are
    returned (customer-specific, route-specific, and every product default —
    including every KG bracket) so the user picks explicitly rather than the
    server silently picking one via priority order.
    """
    id: UUID  # id of the underlying charge_rule or product_charge_option row
    source: str  # "customer" | "route" | "product_default"
    charge_type: ChargeType
    kg_variant: Optional[float] = None
    rate: float


# ---------- Trip ----------

class TripBase(BaseModel):
    driver_id: UUID
    vehicle_id: UUID
    start_date: date
    end_date: Optional[date] = None
    status: TripStatus = TripStatus.ongoing


class TripCreate(TripBase):
    pass


class Trip(TripBase):
    id: UUID
    created_at: datetime


# ---------- Load ----------

class LoadBase(BaseModel):
    trip_id: Optional[UUID] = None  # nullable: supports "unassigned pool" flow
    product_id: UUID
    customer_id: UUID
    origin_place_id: Optional[UUID] = None
    destination_place_id: Optional[UUID] = None
    quantity: Optional[float] = None
    weight_kg: Optional[float] = None
    charge_type: ChargeType
    kg_variant: Optional[float] = None  # which bracket (25/50/100) was selected, if charge_type == kg
    charge: float = 0.0            # resolved or custom-overridden
    charge_rule_used: Optional[str] = None  # e.g. "customer" | "route" | "product_default" | "custom" — for UI display
    discount: float = 0.0          # only meaningful once status == collected
    wages: float = 0.0             # mandatory
    commission_loading: float = 0.0
    commission_unloading: float = 0.0
    status: LoadStatus = LoadStatus.pending


class LoadCreate(LoadBase):
    pass


class LoadCollectRequest(BaseModel):
    """
    Payload for the load-completion step. Wages is asked here (not at load
    creation) — the owner only knows the real wages/commission once the load
    is actually finished. Discount also still only applies once collected.
    """
    discount: float = 0.0
    wages: Optional[float] = None
    commission_loading: Optional[float] = None
    commission_unloading: Optional[float] = None


class Load(LoadBase):
    id: UUID
    created_at: datetime
    net: float = 0.0  # computed: charge - discount - wages - commission


# ---------- Trip Expense ----------

class TripExpenseBase(BaseModel):
    trip_id: UUID
    category: TripExpenseCategory
    custom_label: Optional[str] = None  # used when category == other / custom tile
    amount: float
    note: Optional[str] = None
    date: date


class TripExpenseCreate(TripExpenseBase):
    pass


class TripExpenseUpdate(BaseModel):
    trip_id: Optional[UUID] = None
    category: Optional[TripExpenseCategory] = None
    custom_label: Optional[str] = None
    amount: Optional[float] = None
    note: Optional[str] = None
    date: Optional[date] = None


class TripExpense(TripExpenseBase):
    id: UUID
    created_at: datetime


# ---------- Vehicle Expense ----------

class VehicleExpenseBase(BaseModel):
    vehicle_id: UUID
    category: str
    amount: float
    is_recurring: bool = False
    date: date
    note: Optional[str] = None


class VehicleExpenseCreate(VehicleExpenseBase):
    pass


class VehicleExpense(VehicleExpenseBase):
    id: UUID
    created_at: datetime


# ---------- Driver Settlement ----------

class DriverSettlementBase(BaseModel):
    driver_id: UUID
    trip_id: Optional[UUID] = None  # open item: optionally linked to a trip
    amount: float
    date: date
    note: Optional[str] = None


class DriverSettlementCreate(DriverSettlementBase):
    pass


class DriverSettlement(DriverSettlementBase):
    id: UUID
    created_at: datetime


# ---------- Aggregate / computed response shapes ----------

class DebtSummary(BaseModel):
    collected: float
    pending: float
    expenses: float
    settlements: float
    net_debt: float


class TripDebtSummary(DebtSummary):
    trip_id: UUID


class DriverLedgerEntry(BaseModel):
    date: date
    type: str  # "collection" | "expense" | "settlement"
    label: str
    amount: float  # signed: +collection, -expense, -settlement
    running_balance: float